Shortcodes: Add support for proxying Instagram embeds via WP.com API

ebinnion opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Adds some logic for conditionally proxying embed requests through a new `/oembed-proxy/instagram` endpoint which will then make an authenticated request to Facebook's API to get embed information for a given Instagram URL.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

See p7H4VZ-2DU-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

Yes. 

This PR will begin making authenticated requests to the WordPress.com API to get embed information for Instagram URLs which will create logs for those requests. These logs have the standard retention policy as other API request logs.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

- In WPCOM sandbox, checkout D47837-code
- Be sure to sandbox Jetpack site with `JETPACK__SANDBOX_DOMAIN` constant
- Create new post with Instagram URL embedded
- View post on frontend
- Ensure that Instagram URL is embedded (this should cover the standard Jetpack flow)
- Ensure that unit tests pass (this should cover the WP.com flow)

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Begins proxying requests for Instagram embeds through WordPress.com in preparation for Facebook requiring authentication on all of their endpoints on October 24th.

#### Known issues

At the moment, there is one known issue where the embed doesn't properly preview in Gutenberg. I'm posting a screenshot below. If you happen to review this, and are interested, I'd definitely appreciate any debug assistance. 😄 

<img width="803" alt="Screen Shot 2020-08-11 at 9 18 38 PM" src="https://user-images.githubusercontent.com/1126811/89968179-cd4e2d80-dc18-11ea-991f-1caf769f1f9b.png">

ebinnion committed:
Shortcodes: Add support for proxying embeds via WP.com API

ebinnion committed:
Shortcodes: Fix Instagram tests after adding proxy endpoint

ebinnion commented on modules/shortcodes/instagram.php:
```diff
@@ -177,47 +220,50 @@ function jetpack_instagram_handler( $matches, $atts, $url ) {
+		|| 200 !== wp_remote_retrieve_response_code( $response )
+		|| empty( $response_body )
+	) {
+		return '<!-- instagram error: invalid instagram resource -->';
```
As a note, as part of the refactor, this error message is effectively overridden currently in the calling function currently. That's not a huge deal, but should probably be addressed.

jeherve commented on modules/shortcodes/instagram.php:
```diff
@@ -161,8 +163,49 @@ function jetpack_instagram_handler( $matches, $atts, $url ) {
 
-	$url = esc_url_raw( add_query_arg( $url_args, 'https://api.instagram.com/oembed/' ) );
+	$use_cache     = jetpack_instagram_use_cache( $matches, $atts, $passed_url );
+	$cache_key     = 'oembed_response_body_' . md5( add_query_arg( $url_args, 'https://api.instagram.com/oembed/' ) );
```
Should we add the `esc_url_raw` back in here, so the cache key remains the same as in previous versions?

ebinnion replied:
I'm not sure. I pulled it out because I think it makes it slightly more readable and the existing cached values have a relatively short cache ttl anyways. If you think it's necessary, I can add it back in.

jeherve commented on modules/shortcodes/instagram.php:
```diff
@@ -177,47 +220,50 @@ function jetpack_instagram_handler( $matches, $atts, $url ) {
+ */
+function jetpack_instagram_fetch_embed( $args ) {
+	if ( Constants::is_defined( 'IS_WPCOM' ) && Constants::get_constant( 'IS_WPCOM' ) ) {
+		$url      = esc_url_raw(
+			add_query_arg(
+				$args,
+				'https://api.instagram.com/oembed/'
+			)
```
Maybe we should pass that full URL into the `jetpack_instagram_fetch_embed` function, so we don't have to create that `$url` twice?

ebinnion replied:
Normally, I'd agree. But, we'll need to refactor this slightly for thee October release to 1) point to https://graph.facebook.com/v5.0/instagram_oembed? and 2) allow specifying a token via a filter (WPCOM, VIP, etc.). So, in this case, it's short-lived.

jeherve commented on modules/shortcodes/instagram.php:
```diff
@@ -177,47 +220,50 @@ function jetpack_instagram_handler( $matches, $atts, $url ) {
+		$response = wp_remote_get( $url, array( 'redirection' => 0 ) );
+	} else {
+		if ( ! Jetpack::is_active() ) {
+			return '<!-- instagram error: Jetpack is not connected -->';
```
Maybe we could return just the Instagram wrapped up in a link tag, so something is still displayed on the frontend when the connection breaks?

ebinnion replied:
That's not a bad idea, though it is different from how we were previously handling errored API calls. I'm glad to address that in a follow-up PR and then we can make a decision.

jeherve commented on modules/shortcodes/instagram.php:
```diff
@@ -161,8 +163,49 @@ function jetpack_instagram_handler( $matches, $atts, $url ) {
+		: false;
 
+	if ( ! $response_body ) {
+		$response_body = jetpack_instagram_fetch_embed( $url_args );
+	}
+
+	if ( $use_cache ) {
+		wp_cache_set(
+			$cache_key,
+			$response_body,
+			'instagram_embeds',
+			HOUR_IN_SECONDS + wp_rand( 0, HOUR_IN_SECONDS )
+		);
```
In some cases, `jetpack_instagram_fetch_embed` will not return an embed. Should we avoid caching when that happens, so one does not get to keep an error message for a little while, even after they reconnected their site to WordPress.com for example?

ebinnion replied:
Great point sir! I'll get that fixed up.

ebinnion committed:
Instagram: Refactor to return WP_Error; do not cache errored response; ouput linked url on error

ebinnion commented:
@jeherve - I addressed your feedback in the latest commit. To test the failure case, output a linked URL to Instagram, the simplest method seemed to be to not sandbox the API, which resulted in a 404 response.

jeherve commented on modules/shortcodes/instagram.php:
```diff
@@ -161,8 +163,63 @@ function jetpack_instagram_handler( $matches, $atts, $url ) {
+ * @param string $url The URL that was attempted to embed.
+ */
+function jetpack_instagram_output_errored_embed( $url ) {
+	?>
```
Maybe we could provide more info to folks who can connect / reconnect? Something like:

```suggestion
	if (
		current_user_can( 'jetpack_connect' )
		&& ! ( new Status() )->is_offline_mode()
	) {
		echo '<div class="notice error" style="border-left:5px solid #dc3232;padding:1em;background-color:#f8f9f9;">';
		echo wp_kses(
			sprintf(
				/* translators: URL is a link to the Jetpack dashboard. */
				__( 'Your site is not connected to WordPress.com at the moment. Please <a href="%1$s">connect</a> to start using Instagram embeds.', 'jetpack' ),
				Jetpack::admin_url( 'page=jetpack#/dashboard' )
			),
			array(
				'a' => array(
					'href' => array(),
				),
			)
		);
		echo '</div>';
	}
	?>
```

jeherve replied:
As discussed in Slack, it may be best to just not have any embed when the site is not connected to WordPress.com. 
p1597855236215600-slack-CBG1CP4EN

jeherve commented on modules/shortcodes/instagram.php:
```diff
@@ -14,6 +14,8 @@
 
 use Automattic\Jetpack\Assets;
+use Automattic\Jetpack\Connection\Client;
+use Automattic\Jetpack\Constants;
```
```suggestion
use Automattic\Jetpack\Constants;
use Automattic\Jetpack\Status;
```

This is only needed for my previous suggestion.

ebinnion committed:
Embeds: Return linked URL if Jetpack not connected or in dev mode

ebinnion committed:
Embeds: Use translated strings in WP_Error messages forr Instagram embed

dereksmart reviewed (changes requested):
@ebinnion these seem to be failing tests right now 

```
1) WP_Test_Jetpack_Shortcodes_Instagram::test_shortcode_instagram
Failed asserting that '<a href="https://www.instagram.com/p/BnMO9vRleEx/">https://www.instagram.com/p/BnMO9vRleEx/</a>' contains "<blockquote class="instagram-media" data-instgrm-captioned data-instgrm-permalink="https://www.instagram.com/p/BnMO9vRleEx/".
/tmp/wordpress-master/src/wp-content/plugins/jetpack/tests/php/modules/shortcodes/test-class.instagram.php:125
```

ebinnion committed:
Embed: Fix broken Instagram tests by checking for active connection later

dereksmart reviewed (approved):
Seems to work as expected, and I found no regressions. 

@ebinnion I could not reproduce that bug with the embeds block. I even manually typed the same URL in the screenshot, and it embedded properly in the editor view. 

Speaking of the editor view, I was seeing the proxy request being sent from the backend post editor as well. If this is expected behavior, the patch is good to go 👍

jeherve commented:
Merging this for now. I'll let you handle the WordPress.com counterpart whenever you're ready!

ebinnion commented:
Deployed wpcom in r212586-wpcom
